interop-cats/shared/src/main/scala/zio/interop/catszmanaged.scala

```diff
@@ -160,3 +163,36 @@ private class CatsZManagedSemigroupK[R, E] extends SemigroupK[ZManaged[R, E, ?]]
   override def combineK[A](x: ZManaged[R, E, A], y: ZManaged[R, E, A]): ZManaged[R, E, A] =
     x.orElse(y)
 }
+
+private class CatsZManagedSync[R] extends CatsZManagedMonadError[R, Throwable] with Sync[ZManaged[R, Throwable, ?]] {
+
+  override final def delay[A](thunk: => A): ZManaged[R, Throwable, A] = ZManaged.fromEffect(ZIO.effect(thunk))
+
+  override final def suspend[A](thunk: => zio.ZManaged[R, Throwable, A]): zio.ZManaged[R, Throwable, A] =
+    ZManaged.unwrap(ZIO.effect(thunk))
+
+  override def bracketCase[A, B](
+    acquire: zio.ZManaged[R, Throwable, A]
+  )(
+    use: A => zio.ZManaged[R, Throwable, B]
+  )(
+    release: (A, cats.effect.ExitCase[Throwable]) => zio.ZManaged[R, Throwable, Unit]
+  ): zio.ZManaged[R, Throwable, B] =
+    ZManaged.fromEffect {
+      ZIO.uninterruptibleMask { restore =>
+        for {
+          res <- acquire.reserve
+          fin = res.release
+          a   <- res.acquire.onError(c => fin(Exit.halt(c)))
+          b <- restore(use(a).use(ZIO.succeed)).run.flatMap { exit =>
```

Comment:
This means the returned `B` is released early, so it's "invalid"/deallocated outside of bracket if it had a finalizer.
The usecase i have for this i is to express "allocate temporary resources for allocation of B", for this B must remain allocated outside of .use. Do you think it would make sense here to run reserve and acquire here, but make the finalizer Managed? (Same as ZManaged.preallocate)

Reply:
Ah I see, that makes sense as well.
I was thinking that typically `B` would have a dependency on `A`, and would be defunct after the release / `A`s finalizer is run. Also wouldn't that imply that we delay running release immediately?

So you would propose something like this (ignoring resource safety / interruption)?
```
    ZManaged {
      Ref.make[List[Exit[_, _] => URIO[R, _]]](Nil).map { finalizers =>
        Reservation(
          for {
            resA <- acquire.reserve
            a    <- resA.acquire
            resB <- use(a).reserve
            b    <- resB.acquire
            _    <- release(a, exitToExitCase(Exit.unit)).use_(ZIO.unit) // when exactly should this be run?
            _    <- resA.release(Exit.unit)
            _    <- finalizers.update(resB.release :: _)
          } yield b,
          exitU =>
            for {
              fs    <- finalizers.get
              exits <- ZIO.foreach(fs)(_(exitU).run)
            } yield Exit.collectAllPar(exits)
        )
      }
    }
```

Reply:
@mschuwalow 
I think it makes sense to run all actions in bracket order (run release __action__ eagerly after __use__ action), but defer ALL finalizers as is normal in ZManaged control flow, e.g.:

```scala
ZManaged.unwrap {
  ZIO.uninterruptibleMask { restore =>
    for {
      resA <- acquire.reserve
      finA = resA.release
      a    <- resA.acquire.onError(c => finA(Exit.halt(c)))

      resB  <- restore(use(a).reserve.uninterruptible).onError(c => finA(Exit.halt(c)))
      finB  = resB.release
      exitB <- restore(resB.acquire).run.onError(c => finB(Exit.halt(c)).ensuring(finA(Exit.halt(c))))

      allocC <- release(a, exitToExitCase(exitB)).preallocate.onError { c =>
                 finB(Exit.halt(c)).ensuring(finA(Exit.halt(c)))
               }
      finC = (e: Exit[Any, Any]) => allocC.reserve.map(_.release(e))
      b <- ZIO.done(exitB).onError { c =>
            finC(Exit.halt(c)).ensuring(finB(Exit.halt(c)).ensuring(finA(Exit.halt(c))))
          }
    } yield ZManaged.reserve(Reservation(ZIO.succeed(b), e => finC(e).ensuring(finB(e).ensuring(finA(e)))))
  }
}
```

Since you used force push, I'm not sure if this isn't equivalent to what you had before, but this seems similar to your initial version using ZManaged.foldCauseM – I now believe this makes the most sense, it both adheres to running `release` eagerly after `use`, but also the values aren't invalidated ahead of time by running their finalizers (if zmanaged finalizers finalize instead of the release action)

Reply:
I'm on board with this! I'll update the pr in a sec

Reply:
@neko-kai 
I've updated the pr. Can you check whether this is what you had in mind? I believe this should be equivalent to the snippet you posted.